source/engine/graphics/api/vulkan/RenderGraph.cpp

```diff
@@ -126,103 +110,186 @@ auto CreateLitPass(const nc::graphics::Device* device, nc::graphics::GpuAllocato
       m_shadowMappingPasses{},
-      m_litPass{CreateLitPass(device, m_gpuAllocator, m_swapchain, m_descriptorSets, dimensions)},
+      m_litPass{},
+      m_postProcessImageViews{},
+      m_dummyShadowMap{Attachment(m_device->VkDevice(), m_gpuAllocator, Vector2{1.0f, 1.0f}, true, vk::SampleCountFlagBits::e1, vk::Format::eD16Unorm)},
+      m_onDescriptorSetsChanged{m_shaderBindingManager->OnResourceLayoutChanged().Connect(this, &RenderGraph::SetDescriptorSetLayoutsDirty)},
+      m_onCommitPointLightConnection{registry->OnCommit<PointLight>().Connect([this](graphics::PointLight&){IncrementShadowPassCount();})},
+      m_onRemovePointLightConnection{registry->OnRemove<PointLight>().Connect([this](Entity){DecrementShadowPassCount();})},
       m_dimensions{dimensions},
       m_screenExtent{},
       m_activeShadowMappingPasses{},
-      m_maxLights{maxLights}
+      m_maxLights{maxLights},
+      m_isDescriptorSetLayoutsDirty{std::array<bool, MaxFramesInFlight>{true, true}}
+{
+    for (auto i : std::views::iota(0u, MaxFramesInFlight))
+    {
+        m_litPass.at(i) = CreateLitPass(device, m_gpuAllocator, m_swapchain, dimensions);
+    }
+
+    auto view = m_dummyShadowMap.view.get();
+    m_postProcessImageViews.emplace(PostProcessImageType::ShadowMap, PostProcessViews
+    {
+        .perFrameViews = std::array<std::vector<vk::ImageView>, MaxFramesInFlight>
+        {
+            std::vector<vk::ImageView>(m_maxLights, view),
+            std::vector<vk::ImageView>(m_maxLights, view)
+        }
+    });
+}
+
+void RenderGraph::SinkPostProcessImages()
+{
+    OPTICK_CATEGORY("RenderGraph::SinkPostProcessImages", Optick::Category::Rendering);
+
+    auto* currentFrame = m_frameManager->CurrentFrameContext();
+    auto frameIndex = currentFrame->Index();
+    auto view = m_dummyShadowMap.view.get();
+    m_postProcessImageViews.at(PostProcessImageType::ShadowMap).perFrameViews.at(frameIndex) = std::vector<vk::ImageView>(m_maxLights, view);
+
+    for (auto i : std::views::iota(0u, m_activeShadowMappingPasses))
+    {
+        m_postProcessImageViews.at(PostProcessImageType::ShadowMap).perFrameViews.at(frameIndex).at(i) = m_shadowMappingPasses.at(frameIndex).at(i)->GetAttachmentView(0u);
+    }
+}
+
+auto RenderGraph::GetPostProcessImages(PostProcessImageType imageType) -> std::vector<vk::ImageView>
+{
+    auto* currentFrame = m_frameManager->CurrentFrameContext();
+    return m_postProcessImageViews.at(imageType).perFrameViews.at(currentFrame->Index());
+}
+
+void RenderGraph::CommitResourceLayout()
 {
-    for (auto i : std::views::iota(0u, m_maxLights))
+    auto* currentFrame = m_frameManager->CurrentFrameContext();
+    auto frameIndex = currentFrame->Index();
+
+    if (!m_isDescriptorSetLayoutsDirty.at(frameIndex))
     {
-        m_shadowMappingPasses.push_back(CreateShadowMappingPass(m_device, m_gpuAllocator, m_swapchain, m_dimensions, i));
+        return;
     }
+    m_shadowMappingPasses.at(frameIndex).clear();
```

Comment:
For each of these calls to `at`, I'd suggest doing it once and storing the result in a reference.

Reply:
done